### PR TITLE
PLT-80: Fix "Output doesn't have the minimum required Ada" error while balancing

### PR DESCRIPTION
**Pull request by sjoerdvisscher**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-contract/src/Wallet/Emulator/MultiAgent.hs [hunk 1]
```diff
@@ -302,12 +303,17 @@ emulatorStateInitialDist mp = emulatorStatePool [EmulatorTx tx] where
...
     -- See [Creating wallets with multiple outputs]
-    mkOutputs (key, vl) = mkOutput key <$> splitHeadinto10 (Wallet.splitOffAdaOnlyValue vl)
+    mkOutputs (key, vl) = mkOutput key <$> splitHeadinto10 (splitOffAdaOnlyValue vl)
     splitHeadinto10 []       = []
     splitHeadinto10 (vl:vls) = replicate (fromIntegral count) (Ada.toValue . (`div` count) . Ada.fromValue $ vl) ++ vls
         where
             -- Make sure we don't make the outputs too small
             count = min 10 $ Ada.fromValue vl `div` minAdaTxOut
+    -- | Split value into an ada-only and an non-ada-only value, making sure each has at least minAdaTxOut.
+    splitOffAdaOnlyValue :: Value -> [Value]
```
Comment (koslambrou) [inline plutus-contract/src/Wallet/Emulator/MultiAgent.hs:313]: Why not `... -> (Value, Value)` here? Seems like we always expect a list of 2 values.

**Review by koslambrou — APPROVED** [review]

File: plutus-contract/src/Wallet/Emulator/Wallet.hs [hunk 2]
```diff
@@ -477,36 +474,45 @@ handleBalanceTx utxo utx = do
...
-    pkos = (\v -> Tx.pubKeyTxOut v pk sk) <$> splitOffAdaOnlyValue vl
+calculateTxChanges
+    :: ( Member (Error WAPI.WalletAPIError) effs
+       )
+    => Params
+    -> Address -- ^ The address for the change output
+    -> Map.Map TxOutRef ChainIndexTxOut -- ^ The current wallet's unspent transaction outputs.
+    -> (Value, Value) -- ^ The unbalanced tx's negative and positive balance.
+    -> Eff effs ((Value, [TxIn]), (Value, [TxOut]))
+calculateTxChanges params addr utxos (neg, pos) = do
```
Comment (ak3n) [inline plutus-contract/src/Wallet/Emulator/Wallet.hs:485]: Feels like it would be nice to have a property test here to verify that we always terminate.

**Review by ak3n — APPROVED** [review]
Nice! Thanks for finishing the story with adjusting and `minAdaTxOut`!

**Comment by sjoerdvisscher** [comment]
@MaximilianAlgehed I've applied your changes to Uniswap here, and together with a minAda fix in `checkOwnOutputConstraint`, `prop_Uniswap` now works. But now `"can create a liquidity pool and add liquidity"` fails, and I can't figure out what it fails on.

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher do you have the log from the failure? I can't switch branches right now due to `nix` complications...

**Comment by sjoerdvisscher** [comment]
@MaximilianAlgehed Hydra has finished so you can see it there: https://hydra.iohk.io/build/16951616/nixlog/180

**Pull request merged**